dockerfiles/README.md

```diff
@@ -0,0 +1,116 @@
+Dockerfiles for Skosmos.
+
+## Running with Docker
+
+The following commands will build and tag the image it with `skosmos:test`,
+and run the container. The container name is `skosmos-web`, but you can customize
+the name, port, and other flags as necessary.
+
+```bash
+# NOTE: the container copies the project sources during build, so the
+# context must be the parent directory, i.e. you MUST build the image
+# from the Skosmos source directory, not from $sources/dockerfiles/
+docker build -t skosmos:test . -f dockerfiles/Dockerfile.ubuntu
+docker run -d --rm --name skosmos-web --net=host skosmos:test
+```
+
+Now Skosmos should be available at `http://localhost/`. See the
```

Comment:
Using port 80 on the host is a bit awkward if you already have, say, Apache listening there (which is normally the case for me). Wasn't there a `-p` option for the `docker run` in an earlier version of this README that mapped the container port 80 to something else? I think it would be better to expose this as e.g. port 9090, as is done in the docker-compose setup.

Reply:
Yes, but not if you want the container to access host URLs like localhost:3030 (localhost being the host not container with skosmos)

There are alternatives though

- use a value like fuseki:3030 and ask user to link containers or create a network
- include fuseki in the image (would need 2 Dockerfiles, so docker compose  or AWS etc can deploy separately)
- there is another docker image that does this kind of host-container port mapping (from host to container). I haven't used it, but it'd be 1 more image for the standalone container

Reply:
Ah, now I see - you have to use `--net=host` to make the container share the host network (so that it can access Fuseki on localhost) instead of sitting on a separate bridged network. So the Apache port inside the container (80) has to match what is seen on the outside, there is no room for a mapping layer.

I think this is OK, as long as it's documented - the README has to say that you can't have anything else listening on port 80 (such as Apache).

It should also be possible to switch the Apache in the container to another port by changing the Listen directive, but maybe that's not a great idea since 80 is the expected, standard port number.

I did find a [discussion about accessing services on the host](https://stackoverflow.com/questions/24319662/from-inside-of-a-docker-container-how-do-i-connect-to-the-localhost-of-the-mach) which mentioned the possibility of using `host.docker.internal` but apparently this only works on macOS and Windows, and the most recent versions of Docker on Linux (version [20.10.0 released 2020-12-08](https://docs.docker.com/engine/release-notes/#20101) included https://github.com/moby/moby/pull/40007). So that could be an option as well, but requires a fairly recent Docker installation - the versions that come pre-packaged with distros are usually older.

Reply:
>Ah, now I see - you have to use --net=host to make the container share the host network (so that it can access Fuseki on localhost) instead of sitting on a separate bridged network. So the Apache port inside the container (80) has to match what is seen on the outside, there is no room for a mapping layer.

Exactly.

>I think this is OK, as long as it's documented - the README has to say that you can't have anything else listening on port 80 (such as Apache).

Added the following.

```diff
diff --git a/dockerfiles/README.md b/dockerfiles/README.md
index b09a1705..8d0e0ed5 100644
--- a/dockerfiles/README.md
+++ b/dockerfiles/README.md
@@ -4,7 +4,11 @@ Dockerfiles for Skosmos.
 
 The following commands will build and tag the image it with `skosmos:test`,
 and run the container. The container name is `skosmos-web`, but you can customize
-the name, port, and other flags as necessary.
+the name, and other flags as necessary. The container will listen to port
+`80` at both the host and container since we are using `--net=host` to allow the
+Skosmos application to access Fuseki at `http://localhost:3030`. You are free to
+modify the command line and the `Dockerfile.ubuntu` and configuration files if you
+would like to deploy it differently.
 
     # NOTE: the container copies the project sources during build, so the
     # context must be the parent directory, i.e. you MUST build the image

```

>It should also be possible to switch the Apache in the container to another port by changing the Listen directive, but maybe that's not a great idea since 80 is the expected, standard port number.

I think anyone that wants to deploy Skosmos to AWS/Google/Azure/etc should be able to understand what we did here, and modify the `Dockerfile` and configuration files to match their network/volume/processing requirements.

>I did find a discussion about accessing services on the host which mentioned the possibility of using host.docker.internal but apparently this only works on macOS and Windows, and the most recent versions of Docker on Linux (version 20.10.0 released 2020-12-08 included moby/moby#40007). So that could be an option as well, but requires a fairly recent Docker installation - the versions that come pre-packaged with distros are usually older.

I think we went down the same rabbit hole here :grinning: I tried the `host.docker.internal` but that didn't work, then found an [issue on github](https://github.com/docker/for-linux/issues/264) to add it to Linux Docker. We can review it later after new versions of Docker are released, or if a user or dev has feedback on how to improve it I think.

Thanks @osma!
Bruno